### ISSUE TITLE
Complex Query Question

### ISSUE THREAD

**Issue by mizziness** [spicywebau/craft-neo#498]
Hey there!  I apologize in advance if my question is hard to parse; it's a bit complicated to explain.

I have a Neo field set up like so:

![image](https://user-images.githubusercontent.com/4164335/132034980-59dc83c8-f91e-4e8a-918a-a2a7ebc99fc7.png)

The front-end looks like: 
![image](https://user-images.githubusercontent.com/4164335/132039469-a9cabcf7-0740-4113-a705-732fdccbdee6.png)

I need to build a query in PHP for all entries in a section that have a block called `customSection` that contains a child block called `hypeInline` with a non-empty value.

Thus far, I can get the number of pages that have the Neo block itself: `Entry::find()->siteId(1)->section('pages')->pageSectionBlocksNew(':notempty:')->count()` but I cannot for the life of me figure out how to filter those results down to just the ones I need.

What I am trying to do is detect whether or not an asset is being utilized in one of the pages within one of these `sectionHypeInline` blocks, but the relations inheritance doesn't seem to get me where I need to go.

**Comment by nickdunn** [comment]
> What I am trying to do is detect whether or not an asset is being utilized in one of the pages within one of these sectionHypeInline blocks

Could you come at this from the other side — start with the asset and then see what blocks it's related to?

Something like:

```
$entry = (this is your asset)
$query = NeoBlock::find();
$query->relatedTo = $entry;
$query->allowOwnerRevisions(false);
$query->allowOwnerDrafts(false);
$neoBlocks = $query->all();

$matchingBlocks = [];

foreach ($blocks as $block) {
    try {
        $owner = $block->getOwner();
        // Supertable can be nested in a Neo so find the parent
        if (get_class($owner) === "benf\\neo\\elements\\Block") {
            $owner = $owner->getOwner();
        }
        if($owner) {
            $matchingBlocks[$owner->id] = $owner;
        }
    } catch (InvalidConfigException $e) {

    }
}

$matchingBlocks = array_unique($matchingBlocks);
```

This will get you a list of parent entries where a Neo block contains a field that is related to your asset. I guess you can expand the initial `NeoBlock::find` to narrow it down to a specific type (`sectionHypeInline`) too.

I'm no expert on this, but pieced this together building my own variant of the Related plugin (https://github.com/wrav/related/blob/master/src/services/RelatedService.php) that added Neo support among others.

**Comment by ttempleton** [comment]
This could also be done by checking the Neo blocks on each entry in the section and seeing for which entries the parent and child blocks exist. Unfortunately it can't really be done with a simple entry query, as it isn't possible to use an entry query to check a specific parent/child block setup. The following code should find the entries; it does only check for the existence of the child block, due to the screenshots showing the asset field is a required field on that block, so a check of the asset field value will need to be added if the field is set not to be required in the future.

```php
// Update the handles as necessary
$sectionHandle = 'sectionHandle';
$neoFieldHandle = 'neoFieldHandle';
$parentBlockTypeHandle = 'parentHandle';
$childBlockTypeHandle = 'childHandle';

$entries = \craft\elements\Entry::find()
    ->section($sectionHandle)
    ->with($neoFieldHandle)
    ->all();
$matchingEntries = [];

foreach ($entries as $entry) {
    $blocks = $entry->getEagerLoadedElements($neoFieldHandle);
    $matchFound = false;

    foreach ($blocks as $block) {
        $block->useMemoized($blocks);
    }

    foreach ($blocks as $block) {
        if ($block->level !== 1 || $block->getType()->handle !== $parentBlockTypeHandle) {
            continue;
        }

        foreach ($block->getChildren()->all() as $childBlock) {
            if ($childBlock->getType()->handle === $childBlockTypeHandle) {
                $matchingEntries[] = $entry;
                $matchFound = true;
                break;
            }
        }

        if ($matchFound) {
            break;
        }
    }
}

return $matchingEntries;
```

@nickdunn's solution may be the better option if the `sectionHypeInline` block is only used as a child block of the `customSection` block, though.

**Comment by mizziness** [comment]
My solution ended up being much simpler:

```
use benf\neo\elements\Block as Neo;

private function getInlineUsage(Asset $asset) 
{

    $siteId = \Craft::$app->sites->getCurrentSite()->id;
    $blocks = Neo::find()
        ->siteId( $siteId )
        ->relatedTo( $asset->id )
        ->distinct();

     return $blocks->all();

}
```

Thanks for making this plugin so simple to use :)
